lib/flow/Advection.cpp

```diff
@@ -43,102 +43,146 @@ Advection::CheckReady() const
     {
-        // Check if the particle is inside of the volume.
-        // Also wrap it along periodic dimensions if enabled.
-        if( !velocity->InsideVolumeVelocity( s.back().time, s.back().location ) )
+        auto& s = _streams[streamIdx];
+        size_t numberOfSteps = s.size() - _separatorCount[streamIdx];
+        while(numberOfSteps < maxSteps )
         {
-            auto& p0 = s.back();
-            // Attempt to apply periodicity
-            bool locChanged = false;
-            auto loc = p0.location;
-            for( int i = 0; i < 3; i++ )    // correct coordinates in each periodic dimension
-            {
-                if( _isPeriodic[i] )
+            auto& past0 = s.back();
+            if( past0.IsSpecial() ) // If the last particle is marked "special,"
+                break;              // terminate stream immediately.
+
+            float dt = deltaT;
+            if( s.size() > 2 )  // If there are at least 3 particles in the stream and
+            {                   // neither is a separator, we also adjust *dt*
+                const auto& past1 = s[ s.size()-2 ];
+                const auto& past2 = s[ s.size()-3 ];
+                if( (!past1.IsSpecial()) && (!past2.IsSpecial()) )
                 {
-                    loc[i] = _applyPeriodic( loc[i], _periodicBounds[i][0], _periodicBounds[i][1] );
-                    locChanged = true;
+                    float mindt = deltaT / 20.0f,   maxdt = deltaT * 20.0f;
+                    dt  = past0.time - past1.time;     // step size used by last integration
+                    dt *= _calcAdjustFactor( past2, past1, past0 );
+                    if( dt > 0 )    // integrate forward 
+                        dt  = glm::clamp( dt, mindt, maxdt );
+                    else            // integrate backward
+                        dt  = glm::clamp( dt, maxdt, mindt );
                 }
+           }
+
+           Particle p1;
+           int rv = 0;
+           switch (method) {
+               case ADVECTION_METHOD::EULER:
+                    rv = _advectEuler( velocity, past0, dt, p1 ); break;
+               case ADVECTION_METHOD::RK4:
+                    rv = _advectRK4(   velocity, past0, dt, p1 ); break;
+           }
+
+           if( rv == 0 ){  // Advection successful, keep the new particle.
+                happened = true;
+                s.emplace_back( p1 );
+                numberOfSteps++;
             }
+            else if( rv == MISSING_VAL ) {
+
+                // This is the annoying part: there are multiple possiblities.
+                // 1) past0 is really located at a missing value location;
+                // 2) past0 is inside the volume, but really close to the boundary,
+                //    causing RK4 method to fail;
+                // 3) past0 is not at a missing location, but out of the volume.
+                // Now we need to detect and deal with each of these possibilities.
+
```

Comment:
Please add a comment indicating why periodic boundaries are being handled here, and not by the Grid class itself.

Reply:
A more descriptive comment is added.